azext_iot/iothub/providers/message_endpoint.py

```diff
@@ -0,0 +1,360 @@
+        raise ResourceNotFoundError("No endpoint found.")
+
+    def list(self, endpoint_type: Optional[str] = None):
+        endpoints = self.hub_resource.properties.routing.endpoints
+        if not endpoint_type:
+            return endpoints
+        endpoint_type = endpoint_type.lower()
+        if EndpointType.EventHub.value == endpoint_type:
+            return endpoints.event_hubs
+        if EndpointType.ServiceBusQueue.value == endpoint_type:
+            return endpoints.service_bus_queues
+        if EndpointType.ServiceBusTopic.value == endpoint_type:
+            return endpoints.service_bus_topics
+        if EndpointType.CosmosDBCollection.value == endpoint_type:
+            return endpoints.cosmos_db_sql_collections
+        if EndpointType.AzureStorageContainer.value == endpoint_type:
+            return endpoints.storage_containers
+
+    def delete(self, endpoint_name: Optional[str] = None, endpoint_type: Optional[str] = None):
+        endpoints = self.hub_resource.properties.routing.endpoints
+        if endpoint_type:
+            endpoint_type = endpoint_type.lower()
+        if endpoint_name:
+            endpoint_name = endpoint_name.lower()
+
+            if not endpoint_type or EndpointType.EventHub.value == endpoint_type:
+                endpoints.event_hubs = [e for e in endpoints.event_hubs if e.name.lower() != endpoint_name]
+            if not endpoint_type or EndpointType.ServiceBusQueue.value == endpoint_type:
+                endpoints.service_bus_queues = [e for e in endpoints.service_bus_queues if e.name.lower() != endpoint_name]
+            if not endpoint_type or EndpointType.ServiceBusTopic.value == endpoint_type:
+                endpoints.service_bus_topics = [e for e in endpoints.service_bus_topics if e.name.lower() != endpoint_name]
+            if not endpoint_type or EndpointType.CosmosDBCollection.value == endpoint_type:
+                endpoints.cosmos_db_sql_collections = [
+                    e for e in endpoints.cosmos_db_sql_collections if e.name.lower() != endpoint_name
+                ]
+            if not endpoint_type or EndpointType.AzureStorageContainer.value == endpoint_type:
+                endpoints.storage_containers = [e for e in endpoints.storage_containers if e.name.lower() != endpoint_name]
+
+        elif endpoint_type:
+            if EndpointType.EventHub.value == endpoint_type:
+                endpoints.event_hubs = []
+            elif EndpointType.ServiceBusQueue.value == endpoint_type:
+                endpoints.service_bus_queues = []
+            elif EndpointType.ServiceBusTopic.value == endpoint_type:
+                endpoints.service_bus_topics = []
+            elif EndpointType.CosmosDBCollection.value == endpoint_type:
+                endpoints.cosmos_db_sql_collections = []
+            elif EndpointType.AzureStorageContainer.value == endpoint_type:
+                endpoints.storage_containers = []
+
+        if not endpoint_type and not endpoint_name:
```

Comment:
could be else

Reply:
I think that works...

My understanding of the desired logic is:
- if name - delete a single endpoint with that name
- if type - delete all endpoints of that type
- if no name and no type - delete all endpoints

So the final else would be:
`not endpoint_name and not endpoint_type`, which I think is the desired outcome.

I'm a bit curious about the `endpoint_name and not endpoint_type` scenario - technically this loops through each type of endpoint removing all with the same name, though I believe we've clarified that endpoint names are unique.

I think the end result is the same, but we may be doing some extra work there.

Reply:
I guess we could just add an if found bool to short circuit but then we would have to choose which types go first (so in this case deleting eventhubs will be faster than cosmos db since to get to cosmos, since the if statement logic goes through each type one by one)

or we can leave as is and say everyone suffers

(tbh this prob wont even make a huge difference timing wise since the max number of endpoints in a normal hub is 10)

Reply:
Going to leave as is and let everyone suffer since the max number of endpoints is 10